### DISCUSSION TITLE
Mitt Async

### DISCUSSION THREAD

**Discussion by developit** [developit/mitt#157] [Announcements]
**Q: Why doesn't Mitt support async event handlers?** (from #122)

Asynchrony is expensive when unused, so it can't feasibly be landed in Mitt itself. Also, there isn't an obvious solution for how data should flow between handlers: a function being async implies it should have a return value (since async functions are just functions that return a Promise) - otherwise "async emit" is effectively "sequential emit" or a queue.

However you *can* implement it as an add-on. I'd recommend calling it `emitAsync()` or something, rather than overwriting the normal `emit()` method:

```js
import mitt from 'mitt';

export function mittAsync(all) {
  const inst = mitt(all);
  inst.emitAsync = emitAsync;
  return inst;
}

async function emitAsync(type, e) {
  let handlers = this.all.get(type);
  if (handlers) for (const f of handlers) await f(e);
  let handlers = this.all.get('*');
  if (handlers) for (const f of handlers) await f(type, e);
}

// usage:
const events = mittWithAsync();
events.on('get', async url => fetch(url).then(r=>r.json()));
await events.emit('get', '/foo.json');
```

**Comment by Kikobeats** [comment]
my proposal to maximize asynchronous capabilities:

```js
function emitAsync (type, data) {
  const fns = []
    .concat(this.emitter.all.get('*'))
    .concat(this.emitter.all.get(type))
    .filter(Boolean)
  return Promise.all(fns.map(fn => fn(data)))
}
```
